File: android/src/main/kotlin/ru/kardenvan/flutter_yandex_ads_plus/platform_api/ad_event_dispatcher_facade/InterstitialAdEventDispatcherFacade.kt
package ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher_facade

import ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher.FlutterYandexAdsEventDispatcher

class InterstitialAdEventDispatcherFacade(
    uid: String,
    dispatcher: FlutterYandexAdsEventDispatcher,
): BasicAdEventDispatcherFacade(uid, dispatcher) {
    fun sendAdShownEvent() {
        sendInterstitialAdEvent(Type.ON_AD_SHOWN)
    }

    fun sendAdDismissedEvent() {
        sendInterstitialAdEvent(Type.ON_AD_DISMISSED)
    }

    private fun sendInterstitialAdEvent(
        type: Type,
        parameters: Map<String, Any?>? = null,
    ) {
        super.sendEvent(type = type.value, parameters = parameters)
    }

    private enum class Type constructor(
        val value: String
    ) {
        ON_AD_SHOWN("onAdShown"),
        ON_AD_DISMISSED("onAdDismissed");
    }
}

File: ios/Classes/platform_api/ad_event_dispatcher_facade/InterstitialAdEventDispatcherFacade.swift
//
//  InterstitialAdEventDispatcherFacade.swift
//  flutter_yandex_ads_plus
//
//  Created by apple on 15.01.2023.
//

import Foundation

class InterstitialAdEventDispatcherFacade: BasicAdEventDispatcherFacade {
    func sendOnFailedToAppearEvent(error: Error) {
        sendInterstitialAdEvent(
            type: InterstitialAdEventType.onFailedToAppear,
            parameters: [
                "code": -1,
                "description": error.localizedDescription
            ]
        )
    }

    func sendOnWillAppearEvent() {
        sendInterstitialAdEvent(type: InterstitialAdEventType.onWillAppear)
    }
    
    func sendOnDidAppearEvent() {
        sendInterstitialAdEvent(type: InterstitialAdEventType.onDidAppear)
    }
    
    func sendOnWillDisappearEvent() {
        sendInterstitialAdEvent(
            type: InterstitialAdEventType.onWillDisappear
        )
    }
    
    func sendOnDidDisappearEvent() {
        sendInterstitialAdEvent(
            type: InterstitialAdEventType.onDidDisappear
        )
    }

    private func sendInterstitialAdEvent(
        type: InterstitialAdEventType,
        parameters: [String: Any?]? = nil
    ) {
        sendEvent(type: type.getValue(), parameters: parameters)
    }
    
    enum InterstitialAdEventType {
        case onWillAppear
        case onDidAppear
        case onWillDisappear
        case onDidDisappear
        case onFailedToAppear
        
        func getValue() -> String {
            switch self {
                case .onFailedToAppear:
                    return "onFailedToAppear"
                case .onWillAppear:
                    return "onWillAppear"
                case .onDidAppear:
                    return "onAdShown"
                case .onWillDisappear:
                    return "onWillDisappear"
                case .onDidDisappear:
                    return "onAdDismissed"
            }
        }
    }
}
